bodhi/server/consumers/pungi_wrapper.py

```diff
@@ -0,0 +1,348 @@
+        buildinstall_phase.start()
+
+        # If any of the following three phases fail, we must ensure that
+        # buildinstall is stopped. Otherwise the whole process will hang.
+        try:
+            gather_phase.start()
+            gather_phase.stop()
+
+            extrafiles_phase.start()
+            extrafiles_phase.stop()
+
+            createrepo_phase.start()
+            createrepo_phase.stop()
+
+        finally:
+            buildinstall_phase.stop()
+
+        if not buildinstall_phase.skip():
+            buildinstall_phase.copy_files()
+
+        ostree_phase.start()
+        ostree_phase.stop()
+
+        # PRODUCTIMG phase
+        productimg_phase.start()
+        productimg_phase.stop()
+
+        self.write_repo_metadata()
+
+        # Start all phases for image artifacts
+        pungi.phases.run_all([createiso_phase, liveimages_phase,
+                              image_build_phase, livemedia_phase,
+                              ostree_installer_phase, osbs_phase])
+
+        image_checksum_phase.start()
+        image_checksum_phase.stop()
+
+        pungi.metadata.write_compose_info(self.compose)
+        self.compose.im.dump(self.compose.paths.compose.metadata("images.json"))
+
+        osbs_phase.dump_metadata()
+
+        # TEST phase
+        test_phase.start()
+        test_phase.stop()
+
+        self.compose.write_status("FINISHED")
+
+        self.create_latest_repo_links()
+
+        self.compose.log_info("Compose finished: %s" % self.compose.topdir)
+
+    def load_variants_config(self):
+        """
+        This is a workaround so we dont have to provide a path of the variants
+        config to pungi. We provide a generated file object which then is injected
+        into pungi compose object.
+        """
+        variants_file_obj = cStringIO.StringIO(self.variants_conf.xml)
+        parser = VariantsXmlParser(variants_file_obj)
+        self.compose.variants = parser.parse()
+        self.compose.all_variants = {}
+        for variant in self.compose.get_variants():
+            self.compose.all_variants[variant.uid] = variant
+        # After the variants object is injected into the pungi compose object
+        # we will write it to disk in the repo.
+        variants_file = self.compose.paths.work.variants_file(arch="global")
+        variants_file_obj.seek(0)
+        # Create create the variants file on disk
+        with open(variants_file, 'w') as fd:
+            shutil.copyfileobj(variants_file_obj, fd)
+            variants_file_obj.close()
+
+    def write_repo_metadata(self):
+        # write treeinfo before ISOs are created
+        for variant in self.compose.get_variants():
+            for arch in variant.arches + ["src"]:
+                pungi.metadata.write_tree_info(self.compose, arch, variant)
+
+        # write .discinfo and media.repo before ISOs are created
+        for variant in self.compose.get_variants():
+            for arch in variant.arches + ["src"]:
```

Comment:
Dude, this is the code from from `/usr/bin/pungi-koji` https://pagure.io/pungi/blob/master/f/bin/pungi-koji#_387

Can we leave this be?  It will be on the chopping block just as soon as we can get over this hurdle and get on to writing the api for pungi.

Reply:
Sure.